client_example/25_wrapper/wrapper_basic_gemm.cpp

```diff
@@ -1,5 +1,7 @@
 // SPDX-License-Identifier: MIT
 // Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
+#if(!defined(__HIP_DEVICE_COMPILE__) || defined(__gfx908__) || defined(__gfx90a__) || \
+    defined(__gfx94__))
```

Comment:
@junliume The expected effect is to compile the code for the **_host_** and for **_MI100/200/300_**, right? Looks good (at least for clang 19, see https://clang.llvm.org/docs/AMDGPUSupport.html) except `__gfx94__` which looks suspicious. If you are in doubt too, then you can try this:

```C++
#if(!defined(__HIP_DEVICE_COMPILE__) || (defined(__gfx9__) && !defined(__gfx906__) ||  && !defined(__gfx900__))
```
The above should enable compilation for all gfx9 parts except Vega10/20.

BTW it's a shame that compiler doesn't have the `__HIP_ARCH_*` feature macro that indicates the presence of xDLOPs (MAI) on target, see https://rocm.docs.amd.com/projects/HIP/en/latest/how-to/hip_porting_guide.html#table-of-architecture-properties. Perhaps it's worth requesting it.

Reply:
@junliume Maybe it is worth to create intermediate macros that would enabled cleaner programming, e.g. put into some header this:
```C++
#define TARGET_GFX9   0
#define TARGET_GFX900 0
#define TARGET_GFX906 0
#define TARGET_GFX908 0
#define TARGET_GFX90A 0
#define TARGET_GFX941 0
#define TARGET_GFX942 0
...
#if !defined(__HIP_DEVICE_COMPILE__)
  #define TARGET_HOST 1
#else
  #define TARGET_HOST 0
  #if defined(__gfx9__)
    #undef TARGET_GFX9
    #define TARGET_GFX9 1
  #endif
  ...
  #if defined(__gfx906__)
    #undef TARGET_GFX906
    #define TARGET_GFX906 1
  #endif
  #if defined(__gfx908__)
    #undef TARGET_GFX908
    #define TARGET_GFX908 1
  #endif
  ...
  #if defined(__gfx940__)
    #undef TARGET_GFX940
    #define TARGET_GFX940 1
  #endif
  #if defined(__gfx941__)
    #undef TARGET_GFX941
    #define TARGET_GFX941 1
  #endif
  ...
#endif

// Features:
#define TARGET_GFX94X (TARGET_GFX941 || TARGET_GFX942 || /* enumerate all gfx94s */ )
#define TARGET_GFX_XDLOPS (TARGET_GFX908 || TARGET_GFX90A || TARGET_GFX94X /* add others in the future*/ )
```

Then you can use this like:
```C++
#if TARGET_HOST || (TARGET_XDLOPS && TARGET_GFX9)
  // Enabled for host and for GFX9 XDLOPS targets...
```
which seems much better readable to me.

Reply:
@atamazov found these lines so maybe we should use existing ones there
https://github.com/ROCm/composable_kernel/blob/fix_1371/include/ck/ck.hpp#L55-L74

Reply:
@junliume At first glance, yes. And you may even find some suitable "feature" macros there, like `CK_USE_AMD_MFMA_BF16_1K_OP` or so (but I am still unable to find things like TARGET_GFX_XDLOPS there)))

However, these just _define_, which enforces to use `defined()` everywhere. What I am proposing is _always defining_ to either 0 or 1, which allows for cleaner constructs. Just to mention.